### Probability

#### Overall Shooting Statistics

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.stats import binom
import math

# %matplotlib inline

shot_data = pd.read_csv('./data/stephen_curry_shots_2023.csv')

shot_results = shot_data.value_counts('result')
num_shots = shot_data.index.size
num_shots_made = shot_results[True]
num_shots_missed = shot_results[False]

print('shots: ', num_shots)
print('made: ', num_shots_made)
print('missed: ', num_shots_missed)

make_prob = num_shots_made / num_shots
miss_prob = num_shots_missed / num_shots

print('probability of make: ', make_prob)
print('probability of miss: ', miss_prob)

shot_type = shot_data.value_counts('shot_type')
num_three_pointers = shot_type[3]
three_point_proportion= num_three_pointers / num_shots
num_two_pointers = shot_type[2]
two_point_proportion = num_two_pointers / num_shots

print('number of three-pointers: ', num_three_pointers)
print('number of two-pointers: ', num_two_pointers)
print('proportion of three-pointers: ', three_point_proportion)
print('proportion of two-pointers: ', two_point_proportion)

shots:  1434
made:  703
missed:  731
probability of make:  0.4902370990237099
probability of miss:  0.5097629009762901
number of three-pointers:  786
number of two-pointers:  648
proportion of three-pointers:  0.5481171548117155
proportion of two-pointers:  0.45188284518828453


| | |
| :--- | :--- |
| shots | 1434 |
| made | 703 |
| missed | 731 |
| three pointers | 786 |
| two pointers | 648 |


|  | calculation | result |
| :--- | :--- | :--- |
| prob of make | $$\frac{made}{shots}$$ | 0.490 |
| prob of miss | $$\frac{missed}{shots}$$ | 0.510 |
| proportion of 2 pointers | $$\frac{two pointers}{shots}$$ | 0.452 |
| proportion of 3 pointers | $$\frac{three pointers}{shots}$$ | 0.548 |


#### Probabilties


In [2]:
n1 = 4
k1 = 3

print('3 of 4 made shots probability (any possible combination): ', binom.pmf(k1, n1, make_prob), '\n')

n2 = 5
k2 = 4

combinations4_5 = math.comb(n2, k2)
print('possible combinations 4 of 5 three pointers: ', combinations4_5)
print('manual calculation: ', math.factorial(n2) / (math.factorial(k2)*(math.factorial(n2-k2))), '\n')

probability4_5 = math.pow(three_point_proportion, k2) * \
    math.pow((1-three_point_proportion), n2-k2)
print('4 of 5 three pointers probability (any one specific combination): ', probability4_5)

print('4 of 5 three pointers probability (any possible combination): ', binom.pmf(k2, n2, three_point_proportion))
print('manual calculation: ', combinations4_5 * probability4_5)

3 of 4 made shots probability (any possible combination):  0.2402407848582661 

possible combinations 4 of 5 three pointers:  5
manual calculation:  5.0 

4 of 5 three pointers probability (any one specific combination):  0.04078678122184275
4 of 5 three pointers probability (any possible combination):  0.20393390610921378
manual calculation:  0.20393390610921375


| | calculation | result |
| :--- | :--- | :--- |
| prob_make | | 0.490 |
| three_point_proportion | | 0.548 |
| prob of making 3 of next 4 shots | `binom.pmf(3, 4, prob_make)` | 0.240 |
| prob of 4 of 5 next shots being 3 pointers | `binom.pmf(4, 5, three_point_proportion)` | 0.204 |

##### Assumptions
In these calculations we are combining shots made across the entire 2023 season. The means we are assuming shot probabilities do not change over time (indicated by date or game quarter). Additionally, this means 'next shot' is vague. Is it 'next shot' given an specific historical point in the data, or is the 'next shot' not yet recorded, ie it hasn't happened yet?

### Conditional Probabilities - Future

$$
\begin{array}{rcl}
P(makes shot) & = & 0.490 \\
P(\neg{makes shot}) & = & 0.510 \\
\end{array}
$$

If the next shot Steph shoots is a three-pointer...
- what is the probability he makes it?
- what is the probability it was taken while his team had the lead (lead = True)

$$
\begin{array}{rcl}
P(three pointer) & = & 0.548 \\
P(makes shot|three pointer) & = & 0.419 \\
P(\neg{makes shot}|three pointer) & = & 0.581 \\
P(lead|three pointer) & = & 0.494 \\
P(\neg{lead}|three pointer) & = & 0.506
\end{array}
$$

If the next shot Steph shoots is a two-pointer...
- what is the probability he makes it?
- what is the probability it was taken while his team had the lead (lead = True)

$$
\begin{array}{rcl}
P(two pointer) & = & 0.452 \\
P(makes shot|two pointer) & = & 0.577 \\
P(\neg{makes shot}|two pointer) & = & 0.423 \\
P(lead|two pointer) & = & 0 \\
P(\neg{lead}|two pointer) & = & 0
\end{array}
$$

In [3]:
print('three-point proportion/prob:', three_point_proportion)
print('two-point proportion/prob:', two_point_proportion)

threePointShots = shot_data[shot_data.shot_type == 3]
made_threePointShots = threePointShots[threePointShots.result == True]
threePointShotsWhileLeading = threePointShots[threePointShots.lead == True]
print('\nnumber of threePointShots:', threePointShots.index.size)
print('number of made threePointShots:', made_threePointShots.index.size)
threePointShotMakeProb = made_threePointShots.index.size / threePointShots.index.size
print('number of threePointShots while leading:', threePointShotsWhileLeading.index.size, ':', threePointShotsWhileLeading.index.size / threePointShots.index.size)
print('number of threePointShots while not leading:', threePointShots.index.size - threePointShotsWhileLeading.index.size, ':', )
print('threePointShot make prob:', threePointShotMakeProb)
print('threePointShot miss prob:', 1 - threePointShotMakeProb)

twoPointShots = shot_data[shot_data.shot_type == 2]
made_twoPointShots = twoPointShots[twoPointShots.result == True]
twoPointShotsWhileLeading = twoPointShots[twoPointShots.lead == True]
print('\nnumber of twoPointShots:', twoPointShots.index.size)
print('number of made twoPointShots:', made_twoPointShots.index.size)
twoPointShotMakeProb = made_twoPointShots.index.size / twoPointShots.index.size
print('number of twoPointShots while leading:', twoPointShotsWhileLeading.index.size)
print('number of twoPointShots while not leading:', twoPointShots.index.size - twoPointShotsWhileLeading.index.size)
print('twoPointShot make prob:', twoPointShotMakeProb)
print('twoPointShot miss prob:', 1 - twoPointShotMakeProb)



three-point proportion/prob: 0.5481171548117155
two-point proportion/prob: 0.45188284518828453

number of threePointShots: 786
number of made threePointShots: 329
number of threePointShots while leading: 388 : 0.49363867684478374
number of threePointShots while not leading: 398 :
threePointShot make prob: 0.41857506361323155
threePointShot miss prob: 0.5814249363867685

number of twoPointShots: 648
number of made twoPointShots: 374
number of twoPointShots while leading: 334
number of twoPointShots while not leading: 314
twoPointShot make prob: 0.5771604938271605
twoPointShot miss prob: 0.4228395061728395
